MapboxVisionTests/VisionManagerTests.swift

```diff
@@ -0,0 +1,139 @@
+@testable import MapboxVision
+import XCTest
+
+class VisionManagerTests: XCTestCase {
+    var visionManager: VisionManager?
+    var dependencies: VisionDependencies?
+
+    override func setUp() {
+        super.setUp()
+
+        let native = MockNative()
+
+        dependencies = VisionDependencies(
+            native: native,
+            synchronizer: MockSynchronizable(),
+            recorder: MockSessionRecorder(),
+            dataProvider: MockDataProvider(),
+            deviceInfo: DeviceInfoProvider()
+        )
+
+        self.visionManager = VisionManager(dependencies: dependencies!, videoSource: MockVideoSource())
+    }
+
+    func test_AssertOnVisionManagerDeallocWithoutDestroy() {
+        // Description:
+        // We called destroy in destructors of VisionManager, BaseVisionManager, VisionManagerNative and VisionManagerBaseNative,
+        // so VisionManagerBaseNative's destroy method called multiple times, so we tried to destroy already destroyed resources,
+        // which caused an EXC_BAD_ACCESS.
+        //
+        // Nevertheless the logic was sort of right, it should do assertFailure on deinit without destroy was called
+        //
+        // Given
+        // object of VisionManager
+        // When
+        // we deallocate it without destroy
+        // Then
+        // Should be assertion failure
+        MVTAssertAssertionFailure(
+            self.visionManager = nil,
+            "VisionManager should call assertioFailure when we release the object without calling destroy"
+        )
+    }
+
+    func test_NoAssertOnVisionManagerDeallocAfterDestroy() {
+        // Given
+        // object of VisionManager
+
+        // When
+        // we call destroy and release the instance
+        self.visionManager?.destroy()
+
+        // Then
+        // Shouldn't be assertion failure
+        MVTAssertNoAssertionFailure(
+            self.visionManager = nil,
+            "VisionManager should be successfully deallocated after calling destroy and releasing the object"
+        )
+    }
+
+    func test_RecordingPermissions() {
+        // Given
+        // VisionManager with default country
+
+        guard let visionManager = self.visionManager, let recorder = dependencies?.recorder as? MockSessionRecorder else {
+            XCTAssert(false)
+            return
+        }
+
+        // When
+        // We set countries
+
+        visionManager.onCountryUpdated(.USA)
+        visionManager.onCountryUpdated(.UK)
+        visionManager.onCountryUpdated(.china)
+        visionManager.onCountryUpdated(.other)
+        visionManager.onCountryUpdated(.unknown)
+
+        // Then
+        // we expect that visionManager will call SessionRecorder's method `startInternal` for all countries but China and that there
+        // will no be any other calls to SessionRecorder. For China we expect `stop` method to be called
+
+        let expectedActions: [MockSessionRecorder.Action] = [
+            .startInternal, // USA
```

Comment:
Instead of adding comments (`// USA`, `// UK` etc.) probably it would be better to introduce a typed dictionary similar to [country : expectedActionForCountry].
Then you can setup countries in for-case loop and check elements from actionsLog.

WDYT? Does that make sense?

Reply:
I think it's overengineering, we all understand what the test is doing, @chezzdev WDYT?

Reply:
Yeah, the test itself is clear.

My point was about comments. If we really need them in order to map the expectations, then I'd move the logic in the code. If I'm wrong and these comments were introduced just as "reminders", then probably it's ok.

Anyway this is no big deal for us and can be treated as nit :)

Reply:
As for me this test does its job and will be rewritten the moment it stops serving its purpose.